include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,205 @@
+        process_set.SubtractJacobianTerms(state.rate_constants_, Yn, state.jacobian_);
+
+        // subtract the inverse of the time step from the diagonal
+        // TODO: handle vectorized jacobian matrix
+        for(auto& jac : state.jacobian_.AsVector()) {
+          jac *= -1;
+        }
+        for (std::size_t i_block = 0; i_block < state.jacobian_.Size(); ++i_block)
+        {
+          auto jacobian_vector = std::next(state.jacobian_.AsVector().begin(), i_block * state.jacobian_.FlatBlockSize());
+          for (const auto& i_elem : jacobian_diagonal_elements)
+            jacobian_vector[i_elem] -= 1 / H;
+        }
+
+        // We want to solve this equation for a zero
+        // (y_{n+1} - y_n) / H = f(t_{n+1}, y_{n+1})
+
+        // try to find the root by factoring and solving the linear system
+        linear_solver.Factor(state.jacobian_, state.lower_matrix_, state.upper_matrix_, singular);
+
+        auto yn1_iter = Yn1.begin();
+        auto yn_iter = Yn.begin();
+        auto residual_iter = residual.begin();
+        auto forcing_iter = forcing.begin();
+        // forcing_blk in camchem
+        // residual = (Yn1 - Yn) / H - forcing;
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter, ++residual_iter, ++forcing_iter) {
+          *residual_iter = (*yn1_iter - *yn_iter) / H - *forcing_iter;
+        }
+
+        // the result of the linear solver will be stored in temp
+        // this represnts the change in the solution
+        linear_solver.Solve(residual, temp, state.lower_matrix_, state.upper_matrix_);
+
+        // solution_blk in camchem
+        // Yn1 = Yn1 + temp;
+        auto temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          *yn1_iter += *temp_iter;
+        }
+
+        // always make sure the solution is positive regardless of which iteration we are on
+        // remove negatives, accept this solution and continue
+        yn1_iter = Yn1.begin();
+        yn_iter = Yn.begin();
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter) {
+          *yn_iter = *yn1_iter = std::max(0.0, *yn1_iter);
+        }
+
+        if (iterations == 0) {
+          // we always want to accept the first iteration
+          continue;
+        }
+
+        // check for convergence
+        temp_iter = temp.begin();
+        yn1_iter = Yn1.begin();
+
+        converged = true;
+        // convergence happens when the absolute value of the change to the solution
+        // is less than a tolerance times the absolute value of the solution
+        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
+          // changes that are much smaller than the tolerance are negligible and we assume can be accepted
+          if (std::abs(*temp_iter) > small) {
+            if (std::abs(*temp_iter) > tol * std::abs(*yn1_iter)) {
+              converged = false;
+              break;
+            }
+          }
+        }
```

Comment:
I would change it to:
```
        for(; temp_iter != temp.end(); ++temp_iter, ++yn1_iter) {
          if (std::abs(*temp_iter) > small) {
            if (std::abs(*temp_iter) > tol * std::abs(*yn1_iter)) {
              converged = false;
              break;
            }
            else { converged = true; }
          }
          else { converged = true; }
        }
```

Reply:
I don't see how this is better. We only need to find a single species that didn't converge. The else's are superfluous when we set `converged = true;` before the loop